File: src/op/grad/vision.cc
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

/*!
 * \file src/op/grad/vision.cc
 * \brief Declaration of gradients
 */
#include "./grad_utils.h"
#include "raf/pass.h"
#include "raf/ir.h"

namespace raf {
namespace op {
namespace grad {

using namespace raf::ir;

Array<Expr> RoiAlignGrad(const Expr& orig_call, const Array<Expr> orig_args, const Var& y,
                         const Expr& dy) {
  static auto op_dx = Op::Get("raf.op.roi_align_dx");
  const CallNode* call = orig_call.as<CallNode>();
  CHECK(call != nullptr);
  const Expr& data = call->args[0];
  const Expr& rois = call->args[1];
  const Expr& pooled_size = call->args[2];
  const Expr& spatial_scale = call->args[3];
  const Expr& sample_ratio = call->args[4];
  const Expr& layout = call->args[5];
  const Expr& mode = call->args[6];
  return {Call(op_dx, {data, rois, dy, pooled_size, spatial_scale, sample_ratio, layout, mode})};
}

RAF_OP_GRAD("raf.op.roi_align", RoiAlignGrad);

RAF_OP_GRAD("raf.op.non_max_suppression", NoGrads<1>);

}  // namespace grad
}  // namespace op
}  // namespace raf

File: python/raf/_tvm_op/vision.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

# pylint: disable=missing-function-docstring
"""Compute definition and schedules for vision functions."""
from .nn import schedule_generic
from .._lib import _reg
from .._lib import strategy, register_compute
from .._lib import tvm as _tvm

_topi = _tvm.topi  # pylint: disable=invalid-name,no-member

_reg.register_strategy("raf.op.tvm.get_valid_counts", strategy.get_valid_counts_strategy)
_reg.register_strategy("raf.op.tvm.non_max_suppression", strategy.nms_strategy)
_reg.register_strategy("raf.op.tvm.roi_align", strategy.roi_align_strategy)


@register_compute("raf.op.tvm.roi_align_dx")
def roi_align_dx_compute(attrs, inputs, output_type):
    # pylint: disable=unused-argument
    # pylint: disable=invalid-name
    # pylint: disable=unused-variable
    data = inputs[0]
    rois = inputs[1]
    dy = inputs[2]
    pooled_size, spatial_scale, sample_ratio, layout, mode = (
        attrs.pooled_size,
        attrs.spatial_scale,
        attrs.sample_ratio,
        attrs.layout,
        attrs.mode,
    )
    pooled_size = _topi.utils.get_const_tuple(pooled_size)
    mode = bytes(mode, encoding="utf-8")
    if layout == "NCHW":
        R = _topi.vision.roi_align_nchw(data, rois, pooled_size, spatial_scale, mode, sample_ratio)
    else:
        R = _topi.vision.roi_align_nhwc(data, rois, pooled_size, spatial_scale, mode, sample_ratio)
    grads = _tvm.te.gradient(R, [data], head=dy)
    return grads


_reg.register_schedule("raf.op.tvm.roi_align_dx", schedule_generic)
